libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -175,21 +159,20 @@ class _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI move_only_function<_ReturnT(
     requires(!is_same_v<remove_cvref_t<_Func>, move_only_function> && !__is_inplace_type<_Func>::value &&
              __is_callable_from<_Func>)
   _LIBCPP_HIDE_FROM_ABI move_only_function(_Func&& __func) noexcept {
-    using _FuncWraps = __function_wrappers<_Func>;
-    using _UnRefFunc = remove_reference_t<_Func>;
+    using _FuncWraps  = __function_wrappers<_Func>;
+    using _StoredFunc = decay_t<_Func>;
 
-    if constexpr ((is_pointer_v<_UnRefFunc> && is_function_v<remove_pointer_t<_UnRefFunc>>) ||
-                  is_member_function_pointer_v<_UnRefFunc>) {
+    if constexpr ((is_pointer_v<_StoredFunc> && is_function_v<remove_pointer_t<_StoredFunc>>) ||
+                  is_member_function_pointer_v<_StoredFunc>) {
       if (__func != nullptr) {
         __call_ = &_FuncWraps::__call;
-        std::construct_at(reinterpret_cast<_UnRefFunc*>(__buffer_.data()), std::forward<_Func>(__func));
+        std::construct_at(reinterpret_cast<_StoredFunc*>(__buffer_.data()), std::forward<_Func>(__func));
       }
-    } else if constexpr (__is_move_only_function<remove_cvref_t<_Func>>::value) {
-      if (!__func) {
-        __call_    = {};
-        __destroy_ = {};
-      } else {
-        __construct<_Func>(std::forward<_Func>(__func));
+    } else if constexpr (__is_move_only_function<_StoredFunc>::value) {
+      if (__func) {
+        __call_    = std::exchange(__func.__call_, nullptr);
+        __destroy_ = std::exchange(__func.__destroy_, nullptr);
+        __buffer_  = std::exchange(__func.__buffer_, array<byte, __buffer_size_>{});
```

Comment:
Zeroing out `__func.__buffer_` doesn't seem necessary. This could be simple assignment.